Components/Framework/src/Pseudospectral/Utils.cpp

```diff
@@ -0,0 +1,315 @@
+
+      using namespace QuICC::Memory;
+      tempOnHostMemorySpace ConverterP(ptr, TransferMode::write | TransferMode::block);
+
+      std::uint32_t cumLayerSize = 0;
+      std::uint32_t layerCounter = 0;
+      ptr[0] = 0;
+      for(std::uint32_t l = 0; l < maxLayers; ++l)
+      {
+         std::uint32_t layerSize = 0;
+         if (layerCounter < nLayers)
+         {
+            auto layerIndex = static_cast<std::uint32_t>(res.idx<QuICC::Dimensions::Data::DAT3D>(layerCounter));
+            if (l == layerIndex)
+            {
+               layerSize = res.dim<QuICC::Dimensions::Data::DAT2D>(layerCounter);
+               ++layerCounter;
+            }
+         }
+         ptr[l+1] = ptr[l] + layerSize;
+         cumLayerSize += layerSize;
+      }
+
+      // idx
+      Memory::MemBlock<std::uint32_t> idxBlock(cumLayerSize, mem.get());
+      View::ViewBase<std::uint32_t> idx(idxBlock.data(), idxBlock.size());
+
+      tempOnHostMemorySpace ConverterI(idx, TransferMode::write);
+
+      std::uint32_t l = 0;
+      for(std::uint32_t k = 0; k < nLayers; ++k)
+      {
+         for(int j = 0; j < res.dim<QuICC::Dimensions::Data::DAT2D>(k); ++j)
+         {
+            // auto layerIndex = res.idx<QuICC::Dimensions::Data::DAT3D>(k);
+            auto columnIndex = res.idx<QuICC::Dimensions::Data::DAT2D>(j, k);
+            // auto columnHeight = res.dim<QuICC::Dimensions::Data::DATB1D>(j, k);
+            idx[l] = columnIndex;
+            ++l;
+         }
+      }
+
+      ptrAndIdxBlock ret;
+      ret.ptr = std::move(ptrBlock);
+      ret.idx = std::move(idxBlock);
+      return ret;
+   }
+
+   template <class SCALAROUT, class SCALARIN, class VIEWATT>
+   void copyEig2View(QuICC::View::View<SCALAROUT, VIEWATT> view, const Eigen::Matrix<SCALARIN, -1, -1>& eig, const TransformResolution& res)
+   {
+      throw std::logic_error("trying to copy different types");
+   }
+
+   template <class SCALAR, class VIEWATT>
+   void copyEig2View(QuICC::View::View<SCALAR, VIEWATT> view, const Eigen::Matrix<SCALAR, -1, -1>& eig, const TransformResolution& res)
+   {
+      std ::uint32_t nLayers = res.dim<QuICC::Dimensions::Data::DAT3D>();
+
+      auto pointers = view.pointers()[1];
+
+      using namespace QuICC::Memory;
+
+      /// \todo permanent host buffer
+      View::ViewBase viewb(view.data(), view.size());
+      tempOnHostMemorySpace Converter(viewb, TransferMode::write | TransferMode::block);
+
+      /// \todo permanent transfer / check width only in debug mode
+      tempOnHostMemorySpace ConverterP(pointers, TransferMode::read | TransferMode::block);
+
+
+      /// copy data to view
+      int start = 0;
+      std::int64_t offSet = 0;
+      for(std::uint32_t p = 0; p < nLayers; ++p)
+      {
+         // layer width
+         int cols = res.dim<QuICC::Dimensions::Data::DAT2D>(p);
+         auto widthView = static_cast<std::uint32_t>(cols);
+         // check only in debug mode
```

Comment:
Should this be enable but behind a ifdef DEBUG?

Reply:
cleaned up